### PR TITLE
[ExportVerilog] Add KEEP attr to Vivado array index bug workaround

### PR DESCRIPTION
**Pull request by fabianschuiki**
Rename the `disallowArrayIndexInlining` lowering option to `mitigateVivadoArrayIndexConstPropBug` to more clearly convey the fact that this is intended as a Vivado bug workaround.

Since this is now Vivado-specific, add the `(* keep = "true" *)` SV attribute to the array index wire, which is necessary to properly mitigate the Vivado synthesis issue we were observing.

### CODE REVIEW

File: lib/Conversion/ExportVerilog/PrepareForEmission.cpp [hunk 1]
```diff
@@ -764,6 +764,40 @@ static LogicalResult legalizeHWModule(Block &block,
...
+          wireOp = maybeWireOp;
+          readOp = maybeReadOp;
+        }
+      }
+
+      // Insert a wire and read if necessary.
+      ImplicitLocOpBuilder builder(op.getLoc(), &op);
+      if (!wireOp) {
+        wireOp =
+            builder.create<WireOp>(op.getOperand(1).getType(), "_ARRAY_IDX");
```
Comment (mikeurbach) [inline lib/Conversion/ExportVerilog/PrepareForEmission.cpp:788]: I think the convention is to prefix with _GEN, so maybe use _GEN_ARRAY_IDX so we can still know it starts with _GEN?

**Review by mikeurbach — APPROVED** [review]
I think the contents of the code change are good to go. It sucks to have to do this, especially for one vendor. We really need to get some notion of target triples rather than needing to lump these vendor-specific flags into a global list of lowering options.

File: lib/Conversion/ExportVerilog/PrepareForEmission.cpp (hunk 1 above)
Comment (uenoku) [inline lib/Conversion/ExportVerilog/PrepareForEmission.cpp:788]: I think the op could be in a procedural region. In that case I think we have to use logic+bpassign. 
```
hw.module @Foo(%a: !hw.array<16xi1>, %b : i4) -> () {
  %c1_i4 = hw.constant 1 : i4
  %0 = comb.add %b, %c1_i4 : i4
  sv.initial {
    %1 = hw.array_get %a[%0] : !hw.array<16xi1>, i4
    sv.verbatim "{{0}}" (%1): i1
  }
  hw.output
}
```

Comment (fabianschuiki) [reply lib/Conversion/ExportVerilog/PrepareForEmission.cpp:788]: Great point! Added a fix for that :+1:

**Review by uenoku — APPROVED** [review]
LGTM!

**Pull request merged**